supervisor_types.go

```diff
@@ -136,27 +136,34 @@ type Dimension struct {
 // SpatialDimensionSet is a unique set of druid datasource spatial dimensions.
 type SpatialDimensionSet []SpatialDimension
 
+// DimensionExclusionsSet represents set of excluded dimensions.
+type DimensionExclusionsSet []string
+
 // DimensionsSpec is responsible for configuring Druid's dimensions. They're a
 // set of columns in Druid's data model that can be used for grouping, filtering
 // or applying aggregations.
 // https://druid.apache.org/docs/latest/ingestion/ingestion-spec#dimensionsspec
 type DimensionsSpec struct {
-	Dimensions           DimensionSet        `json:"dimensions,omitempty"`
-	DimansionExclusions  DimensionSet        `json:"dimansionExclusions,omitempty"`
-	SpatialDimensions    SpatialDimensionSet `json:"spatialDimensions,omitempty"`
-	IncludeAllDimensions bool                `json:"includeAllDimensions,omitempty"`
-	UseSchemaDiscovery   bool                `json:"useSchemaDiscovery,omitempty"`
+	Dimensions           DimensionSet           `json:"dimensions,omitempty"`
+	DimensionExclusions  DimensionExclusionsSet `json:"dimensionExclusions,omitempty"`
+	SpatialDimensions    SpatialDimensionSet    `json:"spatialDimensions,omitempty"`
+	IncludeAllDimensions bool                   `json:"includeAllDimensions,omitempty"`
+	UseSchemaDiscovery   bool                   `json:"useSchemaDiscovery,omitempty"`
+}
+type QueryGranularity struct {
+	Type string `json:"type,omitempty"`
 }
 
 // GranularitySpec allows for configuring operations such as data segment
 // partitioning, truncating timestamps, time chunk segmentation or roll-up.
 // https://druid.apache.org/docs/latest/ingestion/ingestion-spec#granularityspec
 type GranularitySpec struct {
-	Type               string   `json:"type"`
-	SegmentGranularity string   `json:"segmentGranularity,omitempty"`
-	QueryGranularity   string   `json:"queryGranularity,omitempty"`
-	Rollup             bool     `json:"rollup,omitempty"`
-	Intervals          []string `json:"intervals,omitempty"`
+	Type               string `json:"type"`
+	SegmentGranularity string `json:"segmentGranularity,omitempty"`
+	// TODO: this field is problematic as depending on value druid returns string or object
```

Comment:
I am not sure that I fully understand the problem but from I was able to figure out it should not be problematic.

> To unmarshal JSON into a value implementing the Unmarshaler interface, Unmarshal calls that value's UnmarshalJSON method, including when the input is a JSON null.

From https://pkg.go.dev/encoding/json#Unmarshal. 

I think we can implement unmarshaler. It can be something along the lines of https://go.dev/play/p/9d0F5a-3_5M

Would this help?

Reply:
Yes. Thanks a lot - I implemented the suggestion for the `QueryGranularity` and `Dimnension` (which representation can also vary).